api/grpcserver/mesh_service.go

```diff
@@ -610,3 +614,68 @@ func (s MeshService) EpochStream(req *pb.EpochStreamRequest, stream pb.MeshServi
 	)
 	return nil
 }
+
+func (s MeshService) MalfeasanceQuery(ctx context.Context, req *pb.MalfeasanceRequest) (*pb.MalfeasanceResponse, error) {
+	var id types.NodeID
+	if len(req.SmesherHex) == 0 {
+		if len(req.Smesher.Id) != types.NodeIDSize {
+			return nil, status.Error(codes.InvalidArgument,
+				fmt.Sprintf("invalid smesher id length (%d), expected (%d)", len(req.Smesher.Id), types.NodeIDSize))
+		}
+		id = types.BytesToNodeID(req.Smesher.Id)
+	} else {
+		parsed, err := hex.DecodeString(req.SmesherHex)
+		if err != nil {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+		id = types.BytesToNodeID(parsed)
+	}
+	proof, err := s.cdb.GetMalfeasanceProof(id)
+	if err != nil && !errors.Is(err, sql.ErrNotFound) {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	return &pb.MalfeasanceResponse{
+		Proof: events.ToMalfeasancePB(id, proof, req.IncludeProof),
+	}, nil
+}
+
+func (s MeshService) MalfeasanceStream(req *pb.MalfeasanceStreamRequest, stream pb.MeshService_MalfeasanceStreamServer) error {
+	sub := events.SubscribeMalfeasance()
+	if sub == nil {
+		return status.Errorf(codes.FailedPrecondition, "event reporting is not enabled")
+	}
+	eventch, fullch := consumeEvents[events.EventMalfeasance](stream.Context(), sub)
+	if err := stream.SendHeader(metadata.MD{}); err != nil {
+		return status.Errorf(codes.Unavailable, "can't send header")
+	}
+
+	// first serve those already existed locally.
+	if err := s.cdb.IterateMalfeasanceProofs(func(id types.NodeID, mp *types.MalfeasanceProof) error {
+		select {
+		case <-stream.Context().Done():
+			return nil
+		default:
+			res := &pb.MalfeasanceStreamResponse{
+				Proof: events.ToMalfeasancePB(id, mp, req.IncludeProof),
+			}
+			return stream.Send(res)
+		}
+	}); err != nil {
+		return status.Error(codes.Internal, err.Error())
+	}
+
+	for {
+		select {
+		case <-stream.Context().Done():
+			return nil
+		case <-fullch:
+			return status.Errorf(codes.Canceled, "buffer is full")
+		case ev := <-eventch:
```

Comment:
Also depending on how many proofs have been stored in the DB already the first one processed might be quite old (and the buffer might be quite full already). Will this return the same malfeasance proof twice then?

Reply:
in the error handling around streaming, the receiver should already be prepared to de-dupe. it's not a problem if they get the same proof twice. it's more of a problem if the a proof is missed and not sent. that's why i started receiving events before iterating db. the code only emit the event after a proof is saved in the db. this order guarantees more that proof won't be missed.